testing/upgradetest/versions.go

```diff
@@ -248,42 +246,73 @@ func findRequiredVersions(sortedParsedVersions []*version.ParsedSemVer, reqs Ver
 	return upgradableVersions, nil
 }
 
-// PreviousMinor returns the previous minor version available for upgrade.
 func PreviousMinor() (*version.ParsedSemVer, error) {
-	versions, err := GetUpgradableVersions()
+	currentVersion := define.Version()
+	upgradeableVersions, err := GetUpgradableVersions()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get upgradable versions: %w", err)
+		return nil, err
 	}
-	current, err := version.ParseVersion(define.Version())
+
+	return previousMinor(currentVersion, upgradeableVersions, false)
+}
+
+func PreviousMinorReleaseOnly() (*version.ParsedSemVer, error) {
+	currentVersion := define.Version()
+	upgradeableVersions, err := GetUpgradableVersions()
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse the current version %s: %w", define.Version(), err)
+		return nil, err
 	}
 
-	// Special case: if we are in the first release (or a patch release of the first release) of a new major (so vX.0.x), we should
-	// return the latest release from the previous major.
-	if current.Minor() == 0 {
-		// Since the current version is the first release of a new major (vX.0.0), there
-		// will be no minor versions in the versions list from the same major (vX). The list
-		// will only contain minors from the previous major (vX-1). Further, since the
-		// version list is sorted in descending order (newer versions first), we can return the
-		// first item from the list as it will be the newest minor of the previous major.
-		for _, v := range versions {
-			if v.Less(*current) {
-				return v, nil
-			}
-		}
+	return previousMinor(currentVersion, upgradeableVersions, true)
+}
 
-		return nil, ErrNoPreviousMinor
+// previousMinor returns the previous minor version available for upgrade from the given list of upgradeable versions.
+//
+// The function attempts to find a suitable previous version using the following strategy:
+//
+//  1. First, it searches for a released previous minor version within the same major version.
+//     This search skips any versions with prerelease tags or build metadata.
+//
+//  2. If no previous minor is found in the same major, and the current version is either:
+//     - A first major release (minor = 0), or
+//     - A prerelease of the first minor (minor = 1 with prerelease)
+//     Then it falls back to searching the previous major version.
+//
+//  3. When falling back to the previous major:
+//     - If fallbackReleaseVersion is true, only released versions are considered
+//     - If fallbackReleaseVersion is false, any version (including prereleases) is considered
+//     - The most recent qualifying version from the previous major is returned
```

Comment:
This (or maybe part of this) should be the documentation of the public functions.